tests/many_to_many/models.py

```diff
@@ -77,3 +77,28 @@ class InheritedArticleA(AbstractArticle):
 
 class InheritedArticleB(AbstractArticle):
     pass
+
+
+class NullableTargetArticle(models.Model):
+    headline = models.CharField(max_length=100)
+    publications = models.ManyToManyField(
+        Publication, through="NullablePublicationThrough"
+    )
+
+
+class NullablePublicationThrough(models.Model):
+    article = models.ForeignKey(NullableTargetArticle, models.CASCADE)
+    publication = models.ForeignKey(Publication, models.CASCADE, null=True)
+
+
+# Manager and models to test if custom managers disable the SQL Optimization in #29725
+class CustomManager(models.Manager):
+    pass
+
+
+class CustomArticle(models.Model):
+    # Same as the Article class above
+    headline = models.CharField(max_length=100)
+    publications = models.ManyToManyField(Publication, name="publications")
+
+    objects = CustomManager()
```

Comment:
This model and manager are unnecessary. `Article` also has a custom manager.

Reply:
`test_custom_default_manager_exists_count` already tests that this optimization is disabled when a manager is defined:
https://github.com/django/django/blob/c991602ce5798385261381025c06698d7fd30dc5/tests/many_to_many/tests.py#L558-L564

Reply:
We can assert queries to make it explicit:
```diff
+        with self.assertNumQueries(2) as queries:
+            self.assertEqual(self.p2.article_set.count(), self.p2.article_set.all().count())
+        self.assertIn("JOIN", queries[0]["sql"])
+        with self.assertNumQueries(2) as queries:
+            self.assertEqual(
+                self.p3.article_set.exists(), self.p3.article_set.all().exists()
+            )
+        self.assertIn("JOIN", queries[0]["sql"])

```